### ISSUE TITLE
Gradients do not exist for variables manually added to wrapper layer

### ISSUE THREAD

**Issue by joaocaldeira** [tensorflow/tensorflow#38607]
**System information** 
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04.3
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: n/a
- TensorFlow installed from (source or binary): binary
- TensorFlow version: v2.1.0-rc2-17-ge5bf8de
- Python version: 3.6.9
- Bazel version (if compiling from source): n/a
- GCC/Compiler version (if compiling from source): n/a 
- CUDA/cuDNN version: 10.2
- GPU model and memory: RTX 2080 TI

**Describe the current behavior**
When using a `Wrapper` layer, parameters added using `self.add_weight` are being ignored by gradients although the outputs depend on those parameters. A regularization term added using `self.add_loss` also depends on the same parameters.

This is visible as the warnings `WARNING:tensorflow:Gradients do not exist for variables ['loss_ignored_3/p_logit:0', 'loss_ignored_4/p_logit:0'] when minimizing the loss.` come up.

This happened when I was trying to port the concrete dropout implementation in https://github.com/yaringal/ConcreteDropout/blob/master/concrete-dropout-keras.ipynb to tensorflow 2.

A minimal example with some of the complexity stripped away and not necessarily sensible math is posted below.

**Describe the expected behavior**
Gradients should exist for those variables, which should be optimized together with the neural network weights.

**Standalone code to reproduce the issue** 
https://colab.research.google.com/drive/1VQHWXn9UtI--aztmrUYuAgMdeOJ5p22u

An earlier version of this problem was posted in https://stackoverflow.com/questions/61164373/loss-added-to-custom-layer-in-tensorflow-2-is-cleared-when-compiling

**Comment by joaocaldeira** [comment]
Operations were being carried out "outside of the gradient tape". Hat tip to Tom Charnock for the explanation. Working version in:
https://colab.research.google.com/drive/19Q3rERNHQMrV9nM_BNp61byyH0_VE0YP

**Comment by axel971** [comment]
Hello Joacoldeira, I got the same issue with the same code of Yarin Gal. Could you please tell me how you fix it ?

**Comment by joaocaldeira** [comment]
You can find a working example of concrete dropout in https://github.com/deepskies/DeeplyUncertain-Public/blob/master/models/cd.py

**Comment by axel971** [comment]
Thank you ! 
Quick question, I saw in the code we have to provide some hyper-paramteres as weight_regularizer and dropout_regularizer. Do you know what is the best way to initialize them ? The author wrote down in this code the weight_regularizer must be l**2 / (\tau * N) and the dropout_regularizer must be 2 / (\tau * N), but I don't to what that corresponds. It's look like also that the author changed the way to initialize these parameters regarding the experimental application. I am confuse!
